### ISSUE TITLE
STRtree Initialization?

### ISSUE THREAD

**Issue by ljwolf** [JuliaGeo/LibGEOS.jl#78]
Hey, I've been toying around with getting an STRTree initialized, and am running into issues. Any guidance would be appreciated.

I've been looking at using the constructors directly with the tools currently in the library:
```julia

using LibGEOS

# Examples from the readme
p1 = readgeom("POLYGON((0 0,1 0,1 1,0 0))")
p2 = readgeom("POLYGON((0 0,1 0,1 1,0 1,0 0))")
p3 = readgeom("POLYGON((2 0,3 0,3 1,2 1,2 0))")

tree = LibGEOS.GEOSSTRtree_create_r(LibGEOS._context.ptr, 3)
LibGEOS.GEOSSTRtree_insert_r(LibGEOS._context.ptr, tree, p1, 1)
```
But this gives an `unsafe_convert(::Type{Ptr{Nothing}}, ::Polygon` error at [l.1318.](https://github.com/JuliaGeo/LibGEOS.jl/blob/master/src/geos_c.jl#L1318).

I'm not sure if I fully understand, but
1. I'd expect `typeof(tree)` to be `LibGEOS.GEOSSTRTree`, or `Ptr{LibGEOS.GEOSSTRTree}`, not `Ptr{Nothing}`?
2. Using `LibGEOS.GEOSSTRtree_create(3)`, you always get a null, but with the `_r` versions, you get a new address every time (which is what I'd expect) 

Thanks, & I hope that's clear.

**Comment by visr** [comment]
I haven't used this before, but it's clearly a part of the API that has been wrapped, but has received no further treatment to offer a more user friendly API. Luckily with a few small modifications it seems to work:
```julia
using LibGEOS

p1 = readgeom("POLYGON((0 0,1 0,1 1,0 0))")
p2 = readgeom("POLYGON((0 0,1 0,1 1,0 1,0 0))")
p3 = readgeom("POLYGON((2 0,3 0,3 1,2 1,2 0))")

tree = LibGEOS.GEOSSTRtree_create_r(LibGEOS._context.ptr, 3)
item = Ref(1)
LibGEOS.GEOSSTRtree_insert_r(LibGEOS._context.ptr, tree, p1.ptr, item)
```

As you can see I had to directly pass the polygon's underlying pointer `p1.ptr`, and the `item` that is passed needs to be some kind of pointer as well, so I used `Ref(1)` instead of `1`.

Regarding your two questions, `LibGEOS.GEOSSTRTree === Nothing`. So they are in fact the same. This can be a be confusing an not very helpful for dispatch, which is the reason some of these `Ptr{Nothing}`s are wrapped into structs.
https://github.com/JuliaGeo/LibGEOS.jl/blob/7ba59a8c14aa8ef61ddf8e413bbc1aa3e2846af9/src/geos_common.jl#L27-L28

I'm not sure about `_r` versus the other API. Has the non thread safe API been deprecated or turned off by default? In any case it's probably best to stick with the `_r` functions.

https://geos.osgeo.org/doxygen/geos__c_8h_source.html
```
   27  *      - Functions ending with _r are thread safe; see details in RFC 3
   28  *        http://trac.osgeo.org/geos/wiki/RFC3.
   29  *        To avoid using by accident non _r functions,
   30  *        define GEOS_USE_ONLY_R_API before including geos_c.h
```

**Comment by ljwolf** [comment]
> As you can see I had to directly pass the polygon's underlying pointer p1.ptr, and the item that is passed needs to be some kind of pointer as well, so I used Ref(1) instead of 1.

I see! I'm not used to dealing with this distinction... I haven't worked a lot with wrapped C directly. Thanks for making it clear! 

> Regarding your two questions, LibGEOS.GEOSSTRTree === Nothing. So they are in fact the same. This can be a be confusing an not very helpful for dispatch, which is the reason some of these Ptr{Nothing}s are wrapped into structs.

Gotcha. And those types are in `geom_types.jl`.

**Comment by ljwolf** [comment]
Your suggestion did work to construct the tree, but I'm not quite sure how to query it... still getting `unsafe_convert` errors with either `LibGEOS.GEOSTouches_r` and `LibGEOS.touches`. 

```julia
julia> LibGEOS.GEOSSTRtree_query_r(LibGEOS._context.ptr, tree, p3.ptr, LibGEOS.GEOSTouches_r, Ref(1))
ERROR: MethodError: no method matching unsafe_convert(::Type{Ptr{Nothing}}, ::typeof(LibGEOS.GEOSTouches_r))
Closest candidates are:
  unsafe_convert(::Type{Ptr{Nothing}}, ::Base.RefValue{T}) where T at refvalue.jl:30
  unsafe_convert(::Type{Ptr{Nothing}}, ::Base.RefArray{T,A,R} where R where A<:(AbstractArray{T,N} where N)) where T at refpointer.jl:90
  unsafe_convert(::Type{Ptr{Nothing}}, ::Base.CFunction) at c.jl:36

```
But, I did manage to make a first stab at a wrapper around the constructors which could be used for bulk querying. 

```julia
using LibGEOS
_context = LibGEOS._context

BaseGeometry = Union{Point, MultiPoint, LineString, MultiLineString, Polygon, MultiPolygon}

mutable struct RTree
    ptr::LibGEOS.GEOSSTRtree

    function RTree(ptr::LibGEOS.GEOSSTRtree)
        tree = new(ptr)
        tree
    end

    function RTree(ary::Array{T,1}) where {T<:BaseGeometry}
        tree = LibGEOS.GEOSSTRtree_create_r(_context.ptr, length(ary))
        for (i, geometry) in enumerate(ary)
            LibGEOS.GEOSSTRtree_insert_r(_context.ptr, tree, envelope(geometry).ptr, Ref(i))
        tree
    end

end
end

```
I'll need to learn a lot more about the internals of Julia ccall mechanics to proceed, I think. Thanks anyway!

**Comment by visr** [comment]
Ah, nice! The error you get is that it cannot convert the julia function to a `Ptr{Nothing}`. What the C library expects is a C compatible function pointer. If you want that, you can use the `cfunction` macro for that, see also these docs and examples:

https://docs.julialang.org/en/latest/base/c/#Base.@cfunction
https://docs.julialang.org/en/latest/manual/calling-c-and-fortran-code/#Creating-C-Compatible-Julia-Function-Pointers-1

https://github.com/JuliaGeo/LibGEOS.jl/blob/269c6b7d62287e81af26cd216f9334363836b629/src/LibGEOS.jl#L49-L51

https://github.com/JuliaGeo/GDAL.jl/blob/34c100bfc07ec5f6d0e9890db79f7f767b66c85d/src/GDAL.jl#L35-L37

**Comment by ljwolf** [comment]
After working with this a few times and trying to elaborate the RTree struct I posted above, I'm routinely getting segmentation faults when trying to extract this into module-level code. I'd presume this happens because  `using LibGEOS` or `import LibGEOS` still means `_context` is unsafe? 

```
signal (11): Segmentation fault
in expression starting at REPL[5]:1
GEOSEnvelope_r at /home/lw17329/.julia/artifacts/8da7bc429c59d77c64bcea500f95eddbff20db7c/lib/libgeos_c.so (unknown line)
GEOSEnvelope_r at /home/lw17329/.julia/packages/LibGEOS/FtvNb/src/geos_c.jl:278 [inlined]
envelope at /home/lw17329/.julia/packages/LibGEOS/FtvNb/src/geos_functions.jl:398 [inlined]
envelope at /home/lw17329/.julia/packages/LibGEOS/FtvNb/src/geos_functions.jl:398
envelope at /home/lw17329/.julia/packages/LibGEOS/FtvNb/src/geos_operations.jl:60
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2159 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2323
jl_apply at /buildworker/worker/package_linux64/build/src/julia.h:1700 [inlined]
do_call at /buildworker/worker/package_linux64/build/src/interpreter.c:369
eval_value at /buildworker/worker/package_linux64/build/src/interpreter.c:458
eval_stmt_value at /buildworker/worker/package_linux64/build/src/interpreter.c:409 [inlined]
eval_body at /buildworker/worker/package_linux64/build/src/interpreter.c:817
jl_interpret_toplevel_thunk at /buildworker/worker/package_linux64/build/src/interpreter.c:911
jl_toplevel_eval_flex at /buildworker/worker/package_linux64/build/src/toplevel.c:819
jl_toplevel_eval_flex at /buildworker/worker/package_linux64/build/src/toplevel.c:769
jl_toplevel_eval_in at /buildworker/worker/package_linux64/build/src/toplevel.c:848
eval at ./boot.jl:331
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2145 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2323
eval_user_input at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.4/REPL/src/REPL.jl:86
macro expansion at /buildworker/worker/package_linux64/build/usr/share/julia/stdlib/v1.4/REPL/src/REPL.jl:118 [inlined]
#26 at ./task.jl:358
_jl_invoke at /buildworker/worker/package_linux64/build/src/gf.c:2145 [inlined]
jl_apply_generic at /buildworker/worker/package_linux64/build/src/gf.c:2323
jl_apply at /buildworker/worker/package_linux64/build/src/julia.h:1700 [inlined]
start_task at /buildworker/worker/package_linux64/build/src/task.c:687
unknown function (ip: (nil))
Allocations: 1862675 (Pool: 1862176; Big: 499); GC: 2
[1]    15501 segmentation fault (core dumped)  julia
```

Anyway, over my head at the moment,

My eventual aim was to update some work I did a while ago on [constructing an adjacency graphs for geometries](https://github.com/ljwolf/SpatialWeights.jl/blob/master/src/weights.jl), which [generally is only performant using some kind of RTree to pre-filter.](http://ljwolf.org/notebooks/perftesting_graphs.html) Having worked with PyGEOS recently & seeing [this](https://discourse.julialang.org/t/moving-to-geometrybasics/40861), I was hoping to get back into building some of these primitives. But, again, seems it's a bit over my head and will remain so for a while.

**Comment by visr** [comment]
> using LibGEOS or import LibGEOS still means _context is unsafe?

Don't think so. It is always initialized on module load: https://github.com/JuliaGeo/LibGEOS.jl/blob/269c6b7d62287e81af26cd216f9334363836b629/src/LibGEOS.jl#L112-L114

But yeah, I know from experience dealing with the segmentation faults when figuring out how to interface with C can be a bit frustrating. If it doesn't have to be GEOS, you can also check out:

https://github.com/alyst/SpatialIndexing.jl
https://github.com/rdeits/RegionTrees.jl
https://github.com/JuliaGeo/LibSpatialIndex.jl

**Comment by maxfreu** [comment]
I think you have to store the envelopes along with the tree. Otherwise, when you exit the constructor, the pointers will be invalid and you get a segfault.

**Comment by visr** [comment]
Cool! I haven't used these trees before, so I'm not sure I can help much. Maybe others have?

Just so you know, there are several alternatives listed here as well: https://github.com/alyst/SpatialIndexing.jl#see-also

Here: https://libgeos.org/doxygen/geos__c_8h.html#a65bbe4c4e9d05889a87498698f1c0af2
It states the item with which the tree should be queried. But that doesn't answer your last question. Although I see [above](https://github.com/JuliaGeo/LibGEOS.jl/issues/78#issuecomment-669326445) I used integers here.

**Comment by maxfreu** [comment]
MVP:

```julia
struct STRtree{T}
    ptr::Ptr{LibGEOS.GEOSSTRtree}  # void pointer to the tree
    items::T  # any geometry for which an envelope can be derived
end

function STRtree(items; nodecapacity=10)
    ctxptr = LibGEOS._context.ptr
    tree = LibGEOS.GEOSSTRtree_create_r(ctxptr, nodecapacity)
    for item in items
        envptr = LibGEOS.envelope(item).ptr
        LibGEOS.GEOSSTRtree_insert_r(ctxptr, tree, envptr, pointer_from_objref(item))
    end
    return STRtree(tree, items)
end

function query(tree::STRtree, geometry::T) where T
    ctxptr = LibGEOS._context.ptr
    matches = T[]

    function callback(item, userdata)::Ptr{Cvoid}
        item_ = unsafe_pointer_to_objref(item)
        userdata_ = unsafe_pointer_to_objref(userdata)
        push!(userdata_, item_)
        return pointer(Cvoid[])
    end

    cfun_callback = @cfunction($callback, Ptr{Cvoid}, (Ptr{Cvoid}, Ptr{Cvoid}))
    LibGEOS.GEOSSTRtree_query_r(ctxptr, tree.ptr, LibGEOS.envelope(geometry).ptr, cfun_callback, pointer_from_objref(matches))
    return matches
end
```

**Comment by evetion** [comment]
Note that `@cfunction($callback, [...]` isn't supported on arm64 architectures, such as the recent Apple M1 devices.

**Comment by maxfreu** [comment]
I can't make it work with e.g. a global default callback and corresponding cfunction -> always segfaults.

**Comment by visr** [comment]
Fixed by @maxfreu in #115
